tensorflow/python/keras/callbacks.py

```diff
@@ -2442,15 +2452,30 @@ class CustomDialect(csv.excel):
           self.csv_file,
           fieldnames=fieldnames,
           dialect=CustomDialect)
-      if self.append_header:
+      # if user wants to log all, then we append_header
+      # at the end of training
+      if self.append_header and not self.log_all:
         self.writer.writeheader()
 
     row_dict = collections.OrderedDict({'epoch': epoch})
-    row_dict.update((key, handle_value(logs[key])) for key in self.keys)
-    self.writer.writerow(row_dict)
-    self.csv_file.flush()
+    row_dict.update((key, handle_value(logs[key]))
+                    for key in self.keys if key in logs)
+    # if user wants to log all, then we write all rows to csv file
+    # at the end of training
+    if not self.log_all:
+      self.writer.writerow(row_dict)
+      self.csv_file.flush()
+    else:
+      self._row_dicts.append(row_dict)
```

Comment:
Thanks for the PR! The argument log_all seems to be a behavior change: now the rows are not necessarily logged at the end of every epoch, even if the keys are present in logs. Am I understanding correctly? If so, is this expected?

Reply:
@rchao Thanks for the review. After looking over this, it's actually incorrect. I've put a new revision, please review that. 

The idea behind this original PR was that we want to log all keys/columns in the csv file to maintain the CSV file format (same # of columns for each row). However, with this csv callback, you only get the columns per epoch. Meaning that if we have validation freq = 3, you will have rows 1 and 2 in the csv file with maybe 3 metrics lets say, and then you'll have row 3 with the original 3 metrics + 2 additional metrics for validation epoch (val acc and val loss). Now what happens is that you have row 1 and 2 with 3 columns, and row 3 with 5 columns. This would be an invalid csv file since not all rows have the same # of columns.

To initially resolve this issue, I thought it would be okay to just keep all the data in-memory and just write everything out at the end. By the end we would know the max amount of columns, and therefore any data rows that didn't have the column, would just insert as empty string. This way we would have a valid csv file at the end.

Problem with this approach:
- Incorrectly assumed data could fit in memory
- If we ever crash before the end of training, all data is gone